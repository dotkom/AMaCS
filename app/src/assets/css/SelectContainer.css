.component > * + * {
  margin-top: 1rem;
}

.selectables {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  margin: -0.8rem;
}

.selectables > * {
  margin: 0.8rem;
}

.prioritizedSwitch {
  display: flex;
  align-items: center;
}

.prioritizedSwitch > * + * {
  margin-left: 0.5rem;
}
