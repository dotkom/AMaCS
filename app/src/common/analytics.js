import ReactGA from 'react-ga'

const debug = process.env.NODE_ENV === 'development';
const trackingId = process.env.SG_GA_TRACKING_ID;

export function initializeAnalytics(history) {
  ReactGA.initialize(trackingId, {
    debug
  });
  history.listen(logPageView);
  logPageView(history.location);
}

export function logPageView(location) {
  if(trackingId) {
    ReactGA.set({ page: location.pathname + location.search });
    ReactGA.pageview(location.pathname + location.search);
  }
}
