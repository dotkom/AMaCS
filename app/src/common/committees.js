import arrkomIcon from '../../../static/arrkom-square.svg';
import bedkomIcon from '../../../static/bedkom-square.svg';
import dotkomIcon from '../../../static/dotkom-square.svg';
import fagkomIcon from '../../../static/fagkom-square.svg';
import prokomIcon from '../../../static/prokom-square.svg';
import trikomIcon from '../../../static/trikom-square.svg';

const committees = new Map([
  ['arrkom', {
    key: 'arrkom',
    name: "Arrkom",
    info: `Arrangementskomitéen jobber for at informatikkstudentene skal ha sosiale
    tilbud gjennom hele året. Vi arrangerer velkjente arrangement som
    immatrikuleringsball, blåtur, Åre-tur, kryssfest, filmkvelder og mye
    mer! Liker du at andre skal ha det kult? Er du en sosial og sprudlende
    person selv? Da er kanskje Arrangementskomitéen stedet for deg!`,
    icon: arrkomIcon,
  }],
  ['bedkom', {
    key: 'bedkom',
    name: "Bedkom",
    info: `Bedriftskomiteens oppgave er å knytte studentene og næringslivet
    sammen. Som medlem vil du være med å arrangere bedriftspresentasjoner
    hvor bedriftene får fortelle om arbeidsmetoder, miljø og prosjekter
    mm. På denne måten vil du opprette en tidlig kontakt med næringslivet
    og potensielle arbeidstakere under studietiden, samtidig som du får et
    innblikk i hvordan bedrifter rekrutterer studenter. Etter en
    bedriftspresentasjon tar gjerne bedriften studentene ut for å mingle
    og bli kjent under et godt måltid. Bedriftskomiteen er også
    knutepunktet til våre samarbeidspartnere i næringslivet ellers.`,
    icon: bedkomIcon,
  }],
  ['dotkom', {
    key: 'dotkom',
    name: "Dotkom",
    info: `Drifts- og utviklingskomiteen har ansvaret for alle linjeforeningens
    datasystemer. Vi leverer tjenester som nettside, intraweb og wiki,
    e-postsystemer, filområder og mye annet spennende. I en stor
    linjeforening er informasjonsflyt og kvalitet viktig. Drifts- og
    utviklingskomiteen prøver hele tiden å lage smarte løsninger som
    holder en høy standard, slik som i en bedrift. Er du interessert i
    programmering, drifting eller teknologi generelt? Vi kan tilby et
    unikt verv i en meget sosial gjeng som driver med IT på høyt nivå.
    Ingen forkunnskaper er påkrevd annet enn stor lærevilje!`,
    icon: dotkomIcon,
  }],
  ['fagkom', {
    key: "fagkom",
    name: "Fagkom",
    info: `Fag- og kurskomiteen tilbyr kunnskap om ny og spennende teknologi,
    gjennom blant annet kurs og foredrag. Vår jobb er å stimulere
    studentenes faglige interesser. Vi oppsøker kompetansen i
    næringslivet, og samarbeider tett med bedriftskomiteen for å komme i
    kontakt med de aktuelle bedriftene. Som medlem av Fag- og kurskomiteen
    vil du få god innsikt i emner utenfor skolen, og samtidig knytte
    tettere bånd til din fremtidige arbeidsgiver.`,
    icon: fagkomIcon,
  }],
  ['prokom', {
    key: "prokom",
    name: "Prokom",
    info: `Profil- og aviskomiteen sørger for at Onlines ansikt utad er
    profesjonelt og konsistent. Alt profileringsmateriale som klær,
    annonser, brev og brosjyrer, plakater, medaljer og andre effekter
    lages av oss. Vi har også en redaksjon som er ansvarlig for
    linjeforeningens tidsskrift, Offline. Har du en designerspire eller en
    skribent i magen? Vi leter etter deg som er interessert i design,
    markedsføring, journalisme eller skriving!`,
    icon: prokomIcon,
  }],
  ['trikom', {
    key: "trikom",
    name: "Trikom",
    info: `Trivselelskomiteen har ansvaret for trivsel blant informatikkstudentene.
    Dette inkluderer å arrangere små arrangementer i skoletiden,
    for eksempel vaffeldager eller andre ting som skaper trivsel.
    Vi er også ansvarlig for drift av kontoret som vil si å sette opp
    kontorvakter, møter, administrere tilganger og passe på at det er tilstrekkelig med mat, drikke, rekvisitter og lignende.
    Så hvis du har en gledesspreder i deg og liker å spre trivsel til dine medelever er trikom komiteen for deg.`,
    icon: trikomIcon,
  }],
]);

export default committees;
