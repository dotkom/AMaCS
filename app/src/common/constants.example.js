


export const OAUTH_SETTINGS = {
  authority: "http://localhost:8000",
  client_id: "...",
  response_type: "token",
  redirect_uri: "http://localhost:8080/auth",
  scope: "authentication.onlineuser.username.read authentication.onlineuser.first_name.read authentication.onlineuser.last_name.read authentication.onlineuser.email.read authentication.onlineuser.is_member.read authentication.onlineuser.field_of_study.read authentication.onlineuser.nickname.read authentication.onlineuser.rfid.read",
  metadata: {
    authorization_endpoint: "http://localhost:8000/sso/o/authorize/",
    userinfo_endpoint: "http://localhost:8000/sso/user/",
    end_session_endpoint: "http://localhost:8000/sso/o/revoke/"
  }
}
