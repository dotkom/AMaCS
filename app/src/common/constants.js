export const OAUTH_SETTINGS = {
  authority: process.env.SG_AUTH_AUTHORITY,
  client_id: process.env.SG_AUTH_CLIENT_ID,
  response_type: process.env.SG_AUTH_RESPONSE_TYPE,
  redirect_uri: process.env.SG_AUTH_REDIRECT_URI,
  scope: process.env.SG_AUTH_SCOPE,
  metadata: {
    authorization_endpoint: process.env.SG_AUTH_ENDPOINT_AUTH,
    userinfo_endpoint: process.env.SG_AUTH_ENDPOINT_USER,
    end_session_endpoint: process.env.SG_AUTH_ENDPOINT_REVOKE,
  }
}

export const API_SETTINGS = {
  host: process.env.SG_APPLICATION_BACKEND,
  application_endpoint: process.env.SG_APPLICATION_ENDPOINT
} 