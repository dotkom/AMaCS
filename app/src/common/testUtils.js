export function createCommitteeObject(
  name='Testkom',
  icon='MockIcon',
  info='Default info'
) {
  return {
    name, icon, info
  };
}
