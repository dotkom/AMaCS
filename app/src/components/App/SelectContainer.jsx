import React from "react";
import PropTypes from 'prop-types';
import _ from "lodash";

import Selectable from "./Selectable"
import SelectedList from "./SelectedList"
import ToggleSwitch from "../ToggleSwitch"

import committeesMap from 'common/committees';
import _s from 'assets/css/SelectContainer.css';

export function updateSelection(selected, committeeName, maxSelected) {
  const selectedIndex = selected.indexOf(committeeName)
  const alreadySelected = selectedIndex !== -1;
  if(alreadySelected) {
    selected = [
      ...selected.slice(0, selectedIndex),
      ...selected.slice(selectedIndex + 1)
    ];
  } else if(selected.length < maxSelected) {
    selected = [...selected, committeeName];
  }
  return selected;
}

class SelectContainer extends React.Component {
  constructor(props) {
    super(props);

    this.maxSelected = 3;
  }

  handleSelect(committeeName) {
    const { selected, onChange } = this.props;
    const newSelection = updateSelection(selected, committeeName, this.maxSelected);
    this.props.onChange(newSelection);
  }

  render() {
    const { selected, ordered, onOrderedChange } = this.props;
    const committees = this.props.committees || committeesMap;
    return (
      <div className={_s.component}>
        <p>Velg komiteene du ønsker å søke ved å klikke på dem{ordered && ' i prioritert rekkefølge'}.</p>
        <div className={_s.selectables}>
          { Array.from(committees).map(([key, committee]) => (
            <Selectable
              key={key}
              onClick={() => this.handleSelect(key)}
              committee={committee}
              selected={selected.some(c => c.toLowerCase() === committee.name.toLowerCase())}
            />
          ))}
        </div>
        <div className={_s.prioritizedSwitch}>
          <ToggleSwitch checked={ordered} onChange={onOrderedChange} />
          <span>Prioritert rekkefølge?</span>
        </div>
        <div className={_s.selectedList}>
          <SelectedList
            ordered={ordered}
            committees={selected.map(committeeName => committees.get(committeeName))}
            totalChoices={this.maxSelected}
            onChange={(committee) => { this.handleSelect(committee)}}
          />
        </div>
      </div>
    );
  }
}

SelectContainer.defaultProps = {
  committees: committeesMap,
  prioritized: false,
};

SelectContainer.propTypes = {
  committees: PropTypes.instanceOf(Map),
  prioritized: PropTypes.bool,
  selected: PropTypes.array.isRequired,
};

export default SelectContainer;
