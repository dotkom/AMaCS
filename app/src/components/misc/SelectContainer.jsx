import React from "react";
import PropTypes from "prop-types";
import { arrayMove } from "react-sortable-hoc";

import Selectable from "./Selectable"
import SelectedList from "./SelectedList"
import ToggleSwitch from "./ToggleSwitch"

import committeesMap from 'common/committees';
import _s from 'assets/css/SelectContainer.scss';

export function updateSelection(selected, committeeName, maxSelected) {
  const selectedIndex = selected.indexOf(committeeName)
  const alreadySelected = selectedIndex !== -1;
  if(alreadySelected) {
    selected = [
      ...selected.slice(0, selectedIndex),
      ...selected.slice(selectedIndex + 1)
    ];
  } else if(selected.length < maxSelected) {
    selected = [...selected, committeeName];
  }
  return selected;
}

class SelectContainer extends React.Component {
  constructor(props) {
    super(props);

    this.maxSelected = 3;
  }

  handleSelect(committeeName) {
    const { selected } = this.props;
    const newSelection = updateSelection(selected, committeeName, this.maxSelected);
    this.props.onChange(newSelection);
  }

  onSortEnd(oldIndex, newIndex) {
    this.props.onChange(arrayMove(this.props.selected, oldIndex, newIndex));
  };

  render() {
    const { selected, ordered, onOrderedChange } = this.props;
    const committees = this.props.committees || committeesMap;
    return (
      <div className={_s.component}>
        <p>Velg komiteene du ønsker å søke ved å klikke på dem.</p>
        <div className={_s.selectables}>
          { Array.from(committees).map(([key, committee]) => (
            <Selectable
              key={key}
              onClick={() => this.handleSelect(key)}
              committee={committee}
              selected={selected.some(c => c.toLowerCase() === committee.name.toLowerCase())}
            />
          ))}
        </div>
        <div className={_s.prioritizedSwitch}>
          <ToggleSwitch checked={ordered} onChange={onOrderedChange} />
          <span>Prioritert rekkefølge?</span>
        </div>
        {ordered && (
          <p>
            Dra komiteene for å prioritere dem.
          </p>
        )}
        <div className={_s.selectedList}>
          <SelectedList
            disabled={!ordered}
            committees={selected.map(committeeName => committees.get(committeeName))}
            totalChoices={this.maxSelected}
            onSortEnd={(oldIndex, newIndex) => this.onSortEnd(oldIndex, newIndex)}
          />
        </div>
      </div>
    );
  }
}

SelectContainer.defaultProps = {
  committees: committeesMap,
  prioritized: false,
};

SelectContainer.propTypes = {
  committees: PropTypes.instanceOf(Map),
  prioritized: PropTypes.bool,
  selected: PropTypes.array.isRequired,
};

export default SelectContainer;
