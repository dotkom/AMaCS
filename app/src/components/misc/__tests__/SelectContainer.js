import React from 'react';
import { shallow } from 'enzyme';

import SelectContainer, { updateSelection } from '../SelectContainer';

import { createCommitteeObject } from 'common/testUtils';

describe('SelectContainer', () => {
  const committees = new Map([
    ['Test0kom', createCommitteeObject('Test0kom')],
    ['Test1kom', createCommitteeObject('Test1kom')],
    ['Test2kom', createCommitteeObject('Test2kom')],
    ['Test3kom', createCommitteeObject('Test3kom')],
  ]);

  it('renders correctly with none selected', () => {
  const selected = [];
    const wrapper = shallow(
      <SelectContainer committees={committees} selected={selected} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with one selected', () => {
  const selected = ['Test1kom'];
    const wrapper = shallow(
      <SelectContainer committees={committees} selected={selected} />
    );
    expect(wrapper).toMatchSnapshot();
  });


  it('renders correctly with two selected', () => {
  const selected = ['Test2kom', 'Test1kom'];
    const wrapper = shallow(
      <SelectContainer committees={committees} selected={selected} />
    );
    expect(wrapper).toMatchSnapshot();
  });


  it('calls onChange when adding committee', () => {
  const selected = ['Test2kom', 'Test1kom'];
  const onChangeMock = jest.fn();
    const wrapper = shallow(
      <SelectContainer
        committees={committees}
        selected={selected}
        onChange={onChangeMock}
      />
    );
    wrapper.find('Selectable').first().simulate('click');
    expect(onChangeMock).toHaveBeenCalledWith(['Test2kom', 'Test1kom', 'Test0kom']);
  });

  it('calls onChange when removing committee', () => {
    const selected = ['Test2kom', 'Test0kom', 'Test1kom'];
    const onChangeMock = jest.fn();
    const wrapper = shallow(
      <SelectContainer
        committees={committees}
        selected={selected}
        onChange={onChangeMock}
      />
    );
    wrapper.find('Selectable').first().simulate('click');
    expect(onChangeMock).toHaveBeenCalledWith(['Test2kom', 'Test1kom']);
  });

  it('calls onChange when removing committee from SelectedList', () => {
    const selected = ['Test2kom', 'Test0kom', 'Test1kom'];
    const onChangeMock = jest.fn();
    const wrapper = shallow(
      <SelectContainer
        committees={committees}
        selected={selected}
        onChange={onChangeMock}
      />
    );
    wrapper.find('SelectedList').first().simulate('change', 'Test0kom');
    expect(onChangeMock).toHaveBeenCalledWith(['Test2kom', 'Test1kom']);
  });
});

describe('updateSelection', () => {
  it('adds selection to when none selected', () => {
    const selected = [];
    const newSelection = updateSelection(selected, 'New', 3);
    expect(newSelection).toEqual(['New']);
  });

  it('adds selection to when 1/3 selected', () => {
    const selected = ['1'];
    const newSelection = updateSelection(selected, '2', 3);
    expect(newSelection).toEqual(['1', '2']);
  });

  it('adds selection to when 2/3 selected', () => {
    const selected = ['1', '2'];
    const newSelection = updateSelection(selected, '3', 3);
    expect(newSelection).toEqual(['1', '2', '3']);
  });

  it('does nothing if 3/3 selected', () => {
    const selected = ['1', '2', '3'];
    const newSelection = updateSelection(selected, '4', 3);
    expect(newSelection).toEqual(selected);
  });

  it('removes item if already selected', () => {
    const selected = ['1', '2', '3'];
    const newSelection = updateSelection(selected, '2', 3);
    expect(newSelection).toEqual(['1', '3']);
  });


  it('returns empty array if removing last item ', () => {
    const selected = ['2'];
    const newSelection = updateSelection(selected, '2', 3);
    expect(newSelection).toEqual([]);
  });
});
