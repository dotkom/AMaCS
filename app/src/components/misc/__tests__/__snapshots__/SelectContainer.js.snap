// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectContainer renders correctly with none selected 1`] = `
<div
  className="component"
>
  <p>
    Velg komiteene du ønsker å søke ved å klikke på dem.
  </p>
  <div
    className="selectables"
  >
    <Selectable
      committee={
        Object {
          "icon": "MockIcon",
          "info": "Default info",
          "key": "test0kom",
          "name": "Test0kom",
        }
      }
      onClick={[Function]}
      selected={false}
      small={false}
    />
    <Selectable
      committee={
        Object {
          "icon": "MockIcon",
          "info": "Default info",
          "key": "test1kom",
          "name": "Test1kom",
        }
      }
      onClick={[Function]}
      selected={false}
      small={false}
    />
    <Selectable
      committee={
        Object {
          "icon": "MockIcon",
          "info": "Default info",
          "key": "test2kom",
          "name": "Test2kom",
        }
      }
      onClick={[Function]}
      selected={false}
      small={false}
    />
    <Selectable
      committee={
        Object {
          "icon": "MockIcon",
          "info": "Default info",
          "key": "test3kom",
          "name": "Test3kom",
        }
      }
      onClick={[Function]}
      selected={false}
      small={false}
    />
  </div>
  <div
    className="prioritizedSwitch"
  >
    <ToggleSwitch
      checked={false}
    />
    <span>
      Prioritert rekkefølge?
    </span>
  </div>
  <div
    className="selectedList"
  >
    <sortableList(SelectedList)
      axis="y"
      committees={Array []}
      disabled={true}
      distance={0}
      getHelperDimensions={[Function]}
      hideSortableGhost={true}
      lockOffset="50%"
      lockToContainerEdges={false}
      onSortEnd={[Function]}
      pressDelay={0}
      pressThreshold={5}
      shouldCancelStart={[Function]}
      totalChoices={3}
      transitionDuration={300}
      useWindowAsScrollContainer={false}
    />
  </div>
</div>
`;

exports[`SelectContainer renders correctly with one selected 1`] = `
<div
  className="component"
>
  <p>
    Velg komiteene du ønsker å søke ved å klikke på dem.
  </p>
  <div
    className="selectables"
  >
    <Selectable
      committee={
        Object {
          "icon": "MockIcon",
          "info": "Default info",
          "key": "test0kom",
          "name": "Test0kom",
        }
      }
      onClick={[Function]}
      selected={false}
      small={false}
    />
    <Selectable
      committee={
        Object {
          "icon": "MockIcon",
          "info": "Default info",
          "key": "test1kom",
          "name": "Test1kom",
        }
      }
      onClick={[Function]}
      selected={true}
      small={false}
    />
    <Selectable
      committee={
        Object {
          "icon": "MockIcon",
          "info": "Default info",
          "key": "test2kom",
          "name": "Test2kom",
        }
      }
      onClick={[Function]}
      selected={false}
      small={false}
    />
    <Selectable
      committee={
        Object {
          "icon": "MockIcon",
          "info": "Default info",
          "key": "test3kom",
          "name": "Test3kom",
        }
      }
      onClick={[Function]}
      selected={false}
      small={false}
    />
  </div>
  <div
    className="prioritizedSwitch"
  >
    <ToggleSwitch
      checked={false}
    />
    <span>
      Prioritert rekkefølge?
    </span>
  </div>
  <div
    className="selectedList"
  >
    <sortableList(SelectedList)
      axis="y"
      committees={
        Array [
          Object {
            "icon": "MockIcon",
            "info": "Default info",
            "key": "test1kom",
            "name": "Test1kom",
          },
        ]
      }
      disabled={true}
      distance={0}
      getHelperDimensions={[Function]}
      hideSortableGhost={true}
      lockOffset="50%"
      lockToContainerEdges={false}
      onSortEnd={[Function]}
      pressDelay={0}
      pressThreshold={5}
      shouldCancelStart={[Function]}
      totalChoices={3}
      transitionDuration={300}
      useWindowAsScrollContainer={false}
    />
  </div>
</div>
`;

exports[`SelectContainer renders correctly with two selected 1`] = `
<div
  className="component"
>
  <p>
    Velg komiteene du ønsker å søke ved å klikke på dem.
  </p>
  <div
    className="selectables"
  >
    <Selectable
      committee={
        Object {
          "icon": "MockIcon",
          "info": "Default info",
          "key": "test0kom",
          "name": "Test0kom",
        }
      }
      onClick={[Function]}
      selected={false}
      small={false}
    />
    <Selectable
      committee={
        Object {
          "icon": "MockIcon",
          "info": "Default info",
          "key": "test1kom",
          "name": "Test1kom",
        }
      }
      onClick={[Function]}
      selected={true}
      small={false}
    />
    <Selectable
      committee={
        Object {
          "icon": "MockIcon",
          "info": "Default info",
          "key": "test2kom",
          "name": "Test2kom",
        }
      }
      onClick={[Function]}
      selected={true}
      small={false}
    />
    <Selectable
      committee={
        Object {
          "icon": "MockIcon",
          "info": "Default info",
          "key": "test3kom",
          "name": "Test3kom",
        }
      }
      onClick={[Function]}
      selected={false}
      small={false}
    />
  </div>
  <div
    className="prioritizedSwitch"
  >
    <ToggleSwitch
      checked={false}
    />
    <span>
      Prioritert rekkefølge?
    </span>
  </div>
  <div
    className="selectedList"
  >
    <sortableList(SelectedList)
      axis="y"
      committees={
        Array [
          Object {
            "icon": "MockIcon",
            "info": "Default info",
            "key": "test2kom",
            "name": "Test2kom",
          },
          Object {
            "icon": "MockIcon",
            "info": "Default info",
            "key": "test1kom",
            "name": "Test1kom",
          },
        ]
      }
      disabled={true}
      distance={0}
      getHelperDimensions={[Function]}
      hideSortableGhost={true}
      lockOffset="50%"
      lockToContainerEdges={false}
      onSortEnd={[Function]}
      pressDelay={0}
      pressThreshold={5}
      shouldCancelStart={[Function]}
      totalChoices={3}
      transitionDuration={300}
      useWindowAsScrollContainer={false}
    />
  </div>
</div>
`;
